om/include/OMR/Om/Value.hpp

```diff
@@ -72,15 +74,21 @@ static constexpr RawValue POINTER = BoxTag::VALUE | KindTag::POINTER;
 static constexpr RawValue VALUE_MASK = ~BoxKindTag::MASK;
 
 /// The canonical NaN is a positive non-signaling NaN. When a NaN double is
-/// stored into a Value, the NaN is canonicalized. This is to prevent us from
-/// reading true NaNs that look like NaN-boxed values.
-static constexpr RawValue CANONICAL_NAN =
-    Infra::Double::SPECIAL_TAG | Infra::Double::NAN_EXTRA_BITS_MASK;
+/// stored into a Value, the NaN is canonicalized.  Doing so ensures that the
+/// NaN is made quiet and unsigned. This is to prevent us from reading true NaNs
+/// that look like NaN-boxed values.
+static constexpr RawValue CANONICAL_NAN = Infra::Double::SPECIAL_TAG |
+                                          Infra::Double::NAN_QUIET_TAG |
+                                          Infra::Double::NAN_EXTRA_BITS_MASK;
+
+static constexpr bool isCanonicalNan(RawValue value) {
+  return (value & CANONICAL_NAN) == value;
+}
 
 /// if value is a NaN, return the corresponding quiet canonical NaN. Note that
 /// currently, the NaN will lose it's sign.
-static constexpr bool canonicalizeNaN(RawValue value) {
-  return Infra::Double::isNaN(value) ? CANONICAL_NAN : value;
+static constexpr RawValue canonicalizeNaN(RawValue value) {
+  return value & CANONICAL_NAN;
```

Comment:
@rwy0717 I'm guessing I should change this to 
```c++
return Infra::Double::isNaN(value) ? CANONICAL_NAN : value;
```
And then we can call canonicalizeNan in `Value::setDouble()`.

Reply:
yep